applications/crossbar/src/modules/cb_modules_util.erl

```diff
@@ -563,29 +564,91 @@ remove_plaintext_password(Context) ->
+    Options = [{'auth_by', cb_context:auth_account_id(Context)}],
+    case knm_number:get(Number, Options) of
+        {'error', 'not_found'} ->
+            validate_number_ownership(Numbers, Unauthorized, Context);
+        {'error', _} ->
+            validate_number_ownership(Numbers, [Number | Unauthorized], Context);
+        _ ->
+            validate_number_ownership(Numbers, Unauthorized, Context)
+    end.
+
 -type assignment_update() :: {ne_binary(), knm_number:knm_number_return()} |
                              {ne_binary(), {'ok', kz_json:object()}} |
                              {ne_binary(), {'error', any()}}.
 -type assignment_updates() :: [assignment_update()].
 
--spec apply_assignment_updates([{ne_binary(), api_binary()}]) ->
+-spec apply_assignment_updates([{ne_binary(), api_binary()}], cb_context:context()) ->
                                       assignment_updates().
-apply_assignment_updates(Updates) ->
-    [maybe_assign_to_port_number(DID, Assign)
+apply_assignment_updates(Updates, Context) ->
+    AccountId = cb_context:account_id(Context),
+    [maybe_assign_to_port_number(DID, Assign, AccountId)
      || {DID, Assign} <- Updates
     ].
 
--spec maybe_assign_to_port_number(ne_binary(), api_binary()) ->
+-spec maybe_assign_to_port_number(ne_binary(), api_binary(), ne_binary()) ->
                                          assignment_update().
-maybe_assign_to_port_number(DID, Assign) ->
+maybe_assign_to_port_number(DID, Assign, AccountId) ->
     Num = knm_converters:normalize(DID),
     case knm_port_request:get(Num) of
         {'error', _} ->
-            {DID, knm_number:assign_to_app(DID, Assign)};
+            maybe_assign_to_app(DID, Assign, AccountId);
         {'ok', JObj} ->
             {DID, knm_port_request:assign_to_app(Num, Assign, JObj)}
     end.
 
+-spec maybe_assign_to_app(ne_binary(), api_binary(), ne_binary()) ->
+                                 assignment_update().
+maybe_assign_to_app(DID, Assign, AccountId) ->
+    case knm_number:get(DID) of
+        {'ok', Number} ->
+            PhoneNumber = knm_number:phone_number(Number),
+            AssignedTo = knm_phone_number:assigned_to(PhoneNumber),
+            maybe_assign_to_app(DID, Assign, AccountId, AssignedTo);
+        {'error', _}=E -> {DID, E};
+        {'dry_run', _, _} -> {DID, {'error', 'dry_run'}}
+    end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Only update used_by (via assign_to_app) when the account being
+%% operated on is the one that the number is assigned to
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec maybe_assign_to_app(ne_binary(), api_binary(), ne_binary(), api_binary()) ->
+                                 assignment_update().
+maybe_assign_to_app(DID, Assign, AccountId, AccountId) ->
+    {DID, knm_number:assign_to_app(DID, Assign)};
+maybe_assign_to_app(DID, _, _, _) ->
+    {DID, {'error', 'unauthorized'}}.
```

Comment:
Not sure why you're doing some verification at all: the knm_number (or knm_numbers) module will do it anyways.

Is this related to the "moving numbers between account keeps app assignment" issue you mentioned?
This BTW sounds like a bug that can be properly resolved in knm_numbers:move/3 by adding a call to `knm_phone_number:set_used_by(_, undefined)` in there.